torch/csrc/jit/tensorexpr/mem_dependency_checker.cpp

```diff
@@ -0,0 +1,1261 @@
+    insertBuffers(outputs_, AccessType::Output);
+  }
+}
+
+void MemDependencyChecker::visit(const Let* v) {
+  const Stmt* last = lastStmt_;
+  lastStmt_ = v;
+
+  IRVisitor::visit(v);
+
+  lastStmt_ = last;
+
+  const Var* var = v->var();
+  if (knownVarBounds_.count(var) != 0) {
+    currentScope_->shadowedVarBounds[var] = knownVarBounds_[var];
+  }
+
+  currentScope_->localVars.insert(var);
+  knownVarBounds_[var] = {v->value(), v->value()};
+}
+
+// Don't support AtomicAdd yet, it's a bit more complex since it's both a read
+// and a write. It's only inserted during Cuda codegen so this should be okay.
+void MemDependencyChecker::visit(const AtomicAdd* v) {
+  throw std::runtime_error("MemDependencyChecker AtomicAdd unimplemented");
+}
+
+void MemDependencyChecker::updateWriteHistory(
+    std::list<BoundRelationship>& writeHistory,
+    const std::shared_ptr<AccessInfo>& info,
+    int latestAccessToClose,
+    bool insert) {
+  bool isWrite = info->isWrite();
+
+  for (auto it = writeHistory.begin(); it != writeHistory.end();) {
+    auto& indexBounds = it->first;
+    std::shared_ptr<AccessInfo> other = it->second;
+    if (info->hasDependency(other)) {
+      ++it;
+      continue;
+    }
+
+    OverlapKind overlap = overlaps(indexBounds, info->bounds());
+
+    if (overlap == NoOverlap) {
+      ++it;
+      continue;
+    }
+
+    // Only writes can close open accesses.
+    if (!isWrite) {
+      info->addDependency(other);
+      other->addDependent(info);
+      ++it;
+      continue;
+    }
+
+    // If we're not closing accesses we can stop here.
+    if ((int)other->id() > latestAccessToClose) {
+      ++it;
+      continue;
+    }
+
+    if (overlap == TotalOverlapB) {
+      // Total overlap is easy - the new access totally replaces the old.
+      it = writeHistory.erase(it);
+    } else {
+      // The new write partially overlaps a previous write. We want to keep
+      // both, but only track the unconvered part of the earlier write.
+
+      // Determine the slices of the earlier bound not covered by info.
+      auto newBounds =
+          subtractIndicesBounds(indexBounds, info->bounds(), overlap);
+
+      // Erase the old slice.
+      it = writeHistory.erase(it);
+
+      // Add all new slices.
+      for (auto& b : newBounds) {
+        it = writeHistory.insert(it, std::make_pair(b, other));
+      }
+      it++;
+    }
+  }
+
+  if (insert && isWrite) {
+    writeHistory.emplace_back(std::make_pair(info->bounds(), info));
+  }
+}
+
+void MemDependencyChecker::mergeScope(
+    const std::shared_ptr<Scope>& child,
+    const std::shared_ptr<Scope>& parent,
+    bool closeOverlapped) {
+  if (child->accesses_.empty()) {
+    return;
+  }
+
+  int latestAccessToClose =
+      closeOverlapped ? (int)child->accesses_.front()->id() : -1;
+
+  // Update dependencies, but don't add new open writes yet.
+  for (auto& info : child->accesses_) {
+    auto& writeHistory = parent->openWrites_[info->var()];
```

Comment:
As in other places, using the indexing operator here makes me nervous. I'd rather use `find` and assert it's found, because if there's a bug in tracking open writes this will fail silently with the indexing operator. Or do you actually mean to open writes as a side-effect?

Reply:
I do mean to open the container as a side effect. This is essentially a defaultdict from var->deque. I'll add a comment here.

Reply:
If we want the side effect everywhere, comments will do for future readers as well, I think.

Reply:
BTW I think more concise is fine too, along the lines of "NB: we want the indexing operator side-effects."